Rejected envelope from backtracking

```
[2022-06-02 10:01:40,705] [WARN] [akka.projection.r2dbc.internal.R2dbcOffsetStore] [] [] [ShoppingAnalyticsService-akka.actor.default-dispatcher-16] - Rejecting unexpected sequence number [1] for pid [ShoppingCart|20220602-081454.1654150494-115], previous sequence number [7]. Offset: TimestampOffset(2022-06-02T06:14:54.854098Z,1970-01-01T00:00:00Z,Map(ShoppingCart|20220602-081454.1654150494-115 -> 1))
[2022-06-02 10:01:40,711] [ERROR] [akka.projection.r2dbc.internal.R2dbcProjectionImpl$R2dbcInternalProjectionState] [akka://ShoppingAnalyticsService@127.0.0.1:3551] [] [ShoppingAnalyticsService-akka.actor.default-dispatcher-3] - [cart-events-ShoppingCart-768-1023] Failed to process envelope with offset [TimestampOffset(2022-06-02T06:14:54.854098Z,1970-01-01T00:00:00Z,Map(ShoppingCart|20220602-081454.1654150494-115 -> 1))]. Projection will stop as defined by recovery strategy.
java.lang.IllegalStateException: Rejected envelope from backtracking, persistenceId [ShoppingCart|20220602-081454.1654150494-115], seqNr [1] due to unexpected sequence number. Please report this issue at https://github.com/akka/akka-persistence-r2dbc
	at akka.projection.r2dbc.internal.R2dbcOffsetStore.isAccepted(R2dbcOffsetStore.scala:673)
	at akka.projection.r2dbc.internal.R2dbcOffsetStore.isAccepted(R2dbcOffsetStore.scala:591)
```

Looking at the code in `isAccepted` it passes the `isDuplicate` check so maybe related to `currentInflight`. It's an `atLeastOnceAsync` projection.

--- a/projection/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala
+++ b/projection/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala
@@ -526,6 +526,11 @@ private[projection] class R2dbcOffsetStore(
           }
         case _ => true
       }
+    if (newInflight.size >= 10000) {
+      throw new IllegalStateException(
+        s"Too many envelopes in-flight [${newInflight.size}]. " +
+        "Please report this issue at https://github.com/akka/akka-persistence-r2dbc")
+    }
     if (!inflight.compareAndSet(currentInflight, newInflight))
       cleanupInflight(newState) // CAS retry, concurrent update of inflight
   }
@@ -671,6 +676,10 @@ private[projection] class R2dbcOffsetStore(
         val ok = seqNr == prevSeqNr + 1
         if (ok) {
           FutureTrue
+        } else if (seqNr <= currentInflight.getOrElse(pid, 0L)) {
+          // currentInFlight contains those that have been processed or about to be processed in Flow,
+          // but offset not saved yet => ok to handle as duplicate
+          FutureFalse
         } else if (recordWithOffset.envelopeLoaded) {
           logUnexpected()
           FutureFalse

--- a/projection/src/test/scala/akka/projection/r2dbc/R2dbcTimestampOffsetProjectionSpec.scala
+++ b/projection/src/test/scala/akka/projection/r2dbc/R2dbcTimestampOffsetProjectionSpec.scala
@@ -260,6 +260,17 @@ class R2dbcTimestampOffsetProjectionSpec
       slice)
   }
 
+  def backtrackingEnvelope(env: EventEnvelope[String]): EventEnvelope[String] =
+    new EventEnvelope[String](
+      env.offset,
+      env.persistenceId,
+      env.sequenceNr,
+      eventOption = None,
+      env.timestamp,
+      env.eventMetadata,
+      env.entityType,
+      env.slice)
+
   def createEnvelopes(pid: Pid, numberOfEvents: Int): immutable.IndexedSeq[EventEnvelope[String]] = {
     (1 to numberOfEvents).map { n =>
       createEnvelope(pid, n, tick().instant(), s"e$n")
@@ -268,6 +279,7 @@ class R2dbcTimestampOffsetProjectionSpec
 
   def createEnvelopesWithDuplicates(pid1: Pid, pid2: Pid): Vector[EventEnvelope[String]] = {
     val startTime = Instant.now()
+
     Vector(
       createEnvelope(pid1, 1, startTime, s"e1-1"),
       createEnvelope(pid1, 2, startTime.plusMillis(1), s"e1-2"),
@@ -1246,6 +1258,8 @@ class R2dbcTimestampOffsetProjectionSpec
       val envelopes = createEnvelopesWithDuplicates(pid1, pid2)
       val sourceProvider = createSourceProvider(envelopes)
       implicit val offsetStore = createOffsetStore(projectionId, sourceProvider)
+
+      info(s"pid1 [$pid1], pid2 [$pid2]")
 
       val flowHandler =
         FlowWithContext[EventEnvelope[String], ProjectionContext]

--- a/projection/src/test/scala/akka/projection/r2dbc/R2dbcTimestampOffsetStoreSpec.scala
+++ b/projection/src/test/scala/akka/projection/r2dbc/R2dbcTimestampOffsetStoreSpec.scala
@@ -91,6 +91,17 @@ class R2dbcTimestampOffsetStoreSpec
       slice)
   }
 
+  def backtrackingEnvelope(env: EventEnvelope[String]): EventEnvelope[String] =
+    new EventEnvelope[String](
+      env.offset,
+      env.persistenceId,
+      env.sequenceNr,
+      eventOption = None,
+      env.timestamp,
+      env.eventMetadata,
+      env.entityType,
+      env.slice)
+
   def createUpdatedDurableState(
       pid: Pid,
       revision: SeqNr,
@@ -346,12 +357,16 @@ class R2dbcTimestampOffsetStoreSpec
       // seqNr 1 is always accepted
       val env1 = createEnvelope("p4", 1L, startTime.plusMillis(1), "e4-1")
       offsetStore.isAccepted(env1).futureValue shouldBe true
+      offsetStore.isAccepted(backtrackingEnvelope(env1)).futureValue shouldBe true
       // but not if already inflight, seqNr 1 was accepted
       offsetStore.addInflight(env1)
-      offsetStore.isAccepted(createEnvelope("p4", 1L, startTime.plusMillis(1), "e4-1")).futureValue shouldBe false
+      val env1Later = createEnvelope("p4", 1L, startTime.plusMillis(1), "e4-1")
+      offsetStore.isAccepted(env1Later).futureValue shouldBe false
+      offsetStore.isAccepted(backtrackingEnvelope(env1Later)).futureValue shouldBe false
       // subsequent seqNr is accepted
       val env2 = createEnvelope("p4", 2L, startTime.plusMillis(2), "e4-2")
       offsetStore.isAccepted(env2).futureValue shouldBe true
+      offsetStore.isAccepted(backtrackingEnvelope(env2)).futureValue shouldBe true
       offsetStore.addInflight(env2)
       // but not when gap
       offsetStore.isAccepted(createEnvelope("p4", 4L, startTime.plusMillis(3), "e4-4")).futureValue shouldBe false
@@ -370,6 +385,10 @@ class R2dbcTimestampOffsetStoreSpec
       offsetStore.addInflight(env3)
       // and then it's not accepted again
       offsetStore.isAccepted(env3).futureValue shouldBe false
+      offsetStore.isAccepted(backtrackingEnvelope(env3)).futureValue shouldBe false
+      // and not when later seqNr is inflight
+      offsetStore.isAccepted(env2).futureValue shouldBe false
+      offsetStore.isAccepted(backtrackingEnvelope(env2)).futureValue shouldBe false
 
       // +1 to known, and then also subsequent are accepted (needed for grouped)
       val env4 = createEnvelope("p3", 6L, startTime.plusMillis(5), "e3-6")

